File: widthOfBinaryTree.py
from collections import deque
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
# Approach 1- BFS     
    def widthOfBinaryTree(self, root): # TC O(n) // SC O(n)
        if not root: return 0
        queue=deque()
        queue.append([root, 0])
        maxDepth=0
        while queue:
            head, headColumnIndex=queue[0]
            size=len(queue)
            #node, columnIndex=None
            for _ in range(size):
                node, columnIndex=queue.popleft()
                if node.left:
                    queue.append([node.left, columnIndex*2])
                if node.right:
                    queue.append([node.right, columnIndex*2+1])
                    
            maxDepth=max(maxDepth, columnIndex-headColumnIndex+1)
        
        return maxDepth

# Approach 2- DFS 
    def widthOfBinaryTree1(self, root): # TC O(n) // SC O(n)
        maxWidth=0 
        colMap={}
        def dfs(root, depth, columnIndex):
            nonlocal maxWidth
            if not root: return 

            if depth not in colMap:
                colMap[depth]=columnIndex

            firstColIndex=colMap[depth]

            maxWidth=max(maxWidth, columnIndex-firstColIndex+1)
            dfs(root.left, depth+1, columnIndex*2)
            dfs(root.right, depth+1, columnIndex*2+1)

        dfs(root, 0, 0)
        return maxWidth 

root=TreeNode(1)
root.left= TreeNode(2);
root.right= TreeNode(3);
root.left.left= TreeNode(5);
root.left.right= TreeNode(3);
root.right.right= TreeNode(9);
s=Solution()
print(s.widthOfBinaryTree(root))
print(s.widthOfBinaryTree1(root))

File: widthOfBinaryTree.java
import java.util.*;
class Pair{
    Node node;
    int columnIndex;
    
    Pair(Node node, int columnIndex){
        this.node=node;
        this.columnIndex=columnIndex;
    }
}

// Approach 1- BFS 
class Solution { // TC O(n) SC O(1)
    public int widthOfBinaryTree(Node root) {
        if(root==null) return 0;
        Queue<Pair> queue=new LinkedList<>();
        queue.add(new Pair(root, 0));
        int maxDepth=Integer.MIN_VALUE;
        while(!queue.isEmpty()){
            Pair head=queue.peek();
            int size=queue.size();
            Pair p=null;
            for(int i=0; i<size; i++){
                p=queue.poll();
                Node currNode=p.node; 
                if(currNode.left!=null){
                    queue.add(new Pair(currNode.left, p.columnIndex*2));
                }
                if(currNode.right!=null){
                    queue.add(new Pair(currNode.right, p.columnIndex*2+1));
                }
            }
            maxDepth=Math.max(maxDepth, p.columnIndex-head.columnIndex+1);
        }
        return maxDepth;
    }
}

// Approach 2- DFS 
class Solution { // TC O(n) SC O(1)
    Map<Integer, Integer> map=new HashMap<>();
    int maxWidth=0;
    
    public void dfs(TreeNode root, int depth, int columnIndex){
        if(root==null) return; 
        
        if(!map.containsKey(depth)){
            map.put(depth, columnIndex);
        }
        int firstColIndex=map.get(depth);
        
        maxWidth=Math.max(maxWidth, columnIndex-firstColIndex+1);
        dfs(root.left, depth+1, columnIndex*2);
        dfs(root.right, depth+1, columnIndex*2+1);
    }
    
    public int widthOfBinaryTree(TreeNode root) {
        dfs(root, 0, 0);
        return maxWidth;
    }
}

class widthOfBinaryTree{
    public static void main(String[] args) {
        BinaryTree bt= new BinaryTree();
        bt.root=new Node(1);
        bt.root.left=new Node(2);
        bt.root.right=new Node(3);
        bt.root.left.left=new Node(5);
        bt.root.left.right=new Node(3);
        bt.root.right.right=new Node(9);
        Solution s=new Solution();
        System.out.println(s.widthOfBinaryTree(bt.root));
        Solution1 s1=new Solution1();
        System.out.println(s1.widthOfBinaryTree(bt.root));
    }
}